hail/src/main/scala/is/hail/expr/types/physical/PCanonicalBaseStruct.scala

```diff
@@ -0,0 +1,223 @@
+  val nMissingBytes: Int = (nMissing + 7) >>> 3
+  val byteOffsets: Array[Long] = new Array[Long](size)
+  override val byteSize: Long = PBaseStruct.getByteSizeAndOffsets(types, nMissingBytes, byteOffsets)
+  override val alignment: Long = PBaseStruct.alignment(types)
+
+
+  def allocate(region: Region): Long = {
+    region.allocate(alignment, byteSize)
+  }
+
+  def allocate(region: Code[Region]): Code[Long] =
+    region.allocate(alignment, byteSize)
+
+  def initialize(structAddress: Long, setMissing: Boolean = false): Unit = {
+    if (allFieldsRequired) {
+      return
+    }
+
+    Region.setMemory(structAddress, nMissingBytes.toLong, if (setMissing) 0xFF.toByte else 0.toByte)
+  }
+
+  def stagedInitialize(structAddress: Code[Long], setMissing: Boolean = false): Code[Unit] = {
+    if (allFieldsRequired) {
+      return Code._empty
+    }
+
+    Region.setMemory(structAddress, const(nMissingBytes.toLong), const(if (setMissing) 0xFF.toByte else 0.toByte))
+  }
+
+  def isFieldDefined(offset: Long, fieldIdx: Int): Boolean =
+    fieldRequired(fieldIdx) || !Region.loadBit(offset, missingIdx(fieldIdx))
+
+  def isFieldMissing(offset: Code[Long], fieldIdx: Int): Code[Boolean] =
+    if (fieldRequired(fieldIdx))
+      false
+    else
+      Region.loadBit(offset, missingIdx(fieldIdx).toLong)
+
+  def setFieldMissing(offset: Long, fieldIdx: Int) {
+    assert(!fieldRequired(fieldIdx))
+    Region.setBit(offset, missingIdx(fieldIdx))
+  }
+
+  def setFieldMissing(offset: Code[Long], fieldIdx: Int): Code[Unit] = {
+    assert(!fieldRequired(fieldIdx))
+    Region.setBit(offset, missingIdx(fieldIdx).toLong)
+  }
+
+  def setFieldPresent(offset: Long, fieldIdx: Int) {
+    if (!fieldRequired(fieldIdx))
+      Region.clearBit(offset, missingIdx(fieldIdx))
+  }
+
+  def setFieldPresent(offset: Code[Long], fieldIdx: Int): Code[Unit] = {
+    if (!fieldRequired(fieldIdx))
+      Region.clearBit(offset, missingIdx(fieldIdx).toLong)
+    else
+      Code._empty
+  }
+
+  def fieldOffset(structAddress: Long, fieldIdx: Int): Long =
+    structAddress + byteOffsets(fieldIdx)
+
+  def fieldOffset(structAddress: Code[Long], fieldIdx: Int): Code[Long] =
+    structAddress + byteOffsets(fieldIdx)
+
+  def loadField(offset: Long, fieldIdx: Int): Long = {
+    val off = fieldOffset(offset, fieldIdx)
+    types(fieldIdx).fundamentalType match {
+      case _: PArray | _: PBinary => Region.loadAddress(off)
+      case _ => off
+    }
+  }
+
+  def loadField(offset: Code[Long], fieldIdx: Int): Code[Long] =
+    loadField(fieldOffset(offset, fieldIdx), types(fieldIdx))
+
+  private def loadField(fieldOffset: Code[Long], fieldType: PType): Code[Long] = {
+    fieldType.fundamentalType match {
+      case _: PArray | _: PBinary => Region.loadAddress(fieldOffset)
+      case _ => fieldOffset
+    }
+  }
+
+  def deepPointerCopy(mb: MethodBuilder, region: Code[Region], dstStructAddress: Code[Long]): Code[Unit] = {
```

Comment:
is it possible to re-order the changes slightly so that loadField, deepPointer copy don't appear as entirely different?

Reply:
yeah, hence the edit :) (I looked over quickly and thought this was in PCanonicalArray)